src/object_store/wrapped.rs

```diff
@@ -33,51 +33,86 @@ pub struct InternalObjectStore {
 
 impl InternalObjectStore {
     pub fn new(inner: Arc<dyn ObjectStore>, config: schema::ObjectStore) -> Self {
-        let root_uri = match config.clone() {
+        let mut root_uri = match config.clone() {
             schema::ObjectStore::Local(Local { data_dir }) => {
                 let canonical_path = StdPath::new(&data_dir).canonicalize().unwrap();
                 Url::from_directory_path(canonical_path).unwrap()
             }
             schema::ObjectStore::InMemory(_) => Url::from_str("memory://").unwrap(),
-            schema::ObjectStore::S3(S3 { bucket, .. }) => {
-                Url::from_str(&format!("s3://{bucket}")).unwrap()
+            schema::ObjectStore::S3(S3 {
+                bucket, endpoint, ..
+            }) => {
+                if let Some(endpoint) = endpoint {
+                    // We're assuming here that the bucket isn't contained in the endpoint itself
+                    Url::from_str(&format!("{endpoint}/{bucket}")).unwrap()
+                } else {
+                    Url::from_str(&format!("s3://{bucket}")).unwrap()
+                }
             }
             schema::ObjectStore::GCS(GCS { bucket, .. }) => {
                 Url::from_str(&format!("gs://{bucket}")).unwrap()
             }
         };
 
+        // If a configured bucket contains a path without a trailing slash add one.
+        if !root_uri.path().ends_with('/') {
+            root_uri.set_path(&format!("{}/", root_uri.path()));
+        }
+
         Self {
             inner,
             root_uri,
             config,
         }
     }
 
+    // Get the table prefix relative to the root of the internal object store.
+    // This is either just a UUID, or potentially UUID prepended by some path.
+    pub fn table_prefix(&self, table_uuid: Uuid) -> Path {
+        match self.config.clone() {
+            schema::ObjectStore::S3(S3 {
+                endpoint: Some(_), ..
+            }) => {
+                // In case when there's an endpoint specified, the bucket is not the host but the
+                // first element in the path, so we need to discard it when creating a prefix.
+                let mut segments =
+                    self.root_uri.path_segments().unwrap().collect::<Vec<_>>();
+
+                // Add the last path segment which is the table UUID
+                let uuid_str = table_uuid.to_string();
+                segments.push(&uuid_str);
+
+                // Remove the first path segment which is actually the bucket
+                segments.remove(0);
+                Path::from(segments.join("/"))
```

Comment:
Maybe having a `prefix` as a separate config field could simplify this logic or avoid having to do this on every query.

Reply:
Yes, in this case it would.